freqtrade/plugins/pairlist/MarketCapPairList.py

```diff
@@ -197,13 +206,24 @@ def filter_pairlist(self, pairlist: list[str], tickers: dict) -> list[str]:
                 pair_format += f":{self._stake_currency.upper()}"
 
             top_marketcap = marketcap_list[: self._max_rank :]
+            markets = self.get_markets_cache()
 
             for mc_pair in top_marketcap:
                 test_pair = f"{mc_pair.upper()}/{pair_format}"
-                if test_pair in pairlist and test_pair not in filtered_pairlist:
-                    filtered_pairlist.append(test_pair)
-                    if len(filtered_pairlist) == self._number_assets:
-                        break
+                test_prefix_1000 = f"1000{test_pair}"  # Binance
+                test_prefix_k = f"k{test_pair}"  # Hyperliquid
+
+                if not any(p.startswith(test_pair) for p in filtered_pairlist):
+                    if any(p.startswith(test_pair) for p in pairlist):
+                        filtered_pairlist.append(test_pair)
+                    elif not any(p.startswith(test_pair) for p in markets):
+                        if any(p.startswith(test_prefix_1000) for p in pairlist):
+                            filtered_pairlist.append(test_prefix_1000)
+                        elif any(p.startswith(test_prefix_k) for p in pairlist):
+                            filtered_pairlist.append(test_prefix_k)
```

Comment:
i think this whole "monster" should be in it's own function (ideally one without side-effects - passing in what's necessary, and returning "something" to add to the final pairlist)

and it'll definitely have to earn a very precise test ... 

This condition is already difficult to read, letalone understand - so i'd want more than 100% confidence that tests catch eventual bugs in it, especially thinking about future changes to this.